Emscripten support

BinBashBanana opened the pull request:
This change allows mGBA to be built for emscripten

endrift commented on include/mgba-util/formatting.h:
```diff
@@ -20,8 +20,8 @@ typedef const char* locale_t;
 int ftostr_l(char* restrict str, size_t size, float f, locale_t locale);
 
-#ifndef HAVE_STRTOF_L
-float strtof_l(const char* restrict str, char** restrict end, locale_t locale);
```
This will break a lot of stuff. I have no idea what you're even using it for here?

endrift commented on src/util/formatting.c:
```diff
@@ -26,8 +26,8 @@ int ftostr_l(char* restrict str, size_t size, float f, locale_t locale) {
 }
 
-#ifndef HAVE_STRTOF_L
-float strtof_l(const char* restrict str, char** restrict end, locale_t locale) {
```
Ditto

endrift commented on src/util/formatting.c:
```diff
@@ -64,7 +64,7 @@ float strtof_u(const char* restrict str, char** restrict end) {
 #else
 	locale_t l = "C";
 #endif
-	float res = strtof_l(str, end, l);
```
Ditto

endrift reviewed (changes requested):
There's already an upstream branch that supports emscripten, it would be better to cherry-pick commits from it instead of duplicating them here.

BinBashBanana commented:
I had to rename it, because otherwise it would conflict with a variable in libretro_emscripten.bc when I tried to link the final file.

endrift commented:
No, you cannot rename it. Instead you should define HAVE_STRTOF_L.

BinBashBanana commented:
I barely even know C, so I wouldn't know how to do that. What is the branch that supports emscripten?

endrift commented:
It's not a C thing, it's just changing this line:
`DEFINES += -std=gnu99 -DHAVE_LOCALE`
to this:
`DEFINES += -std=gnu99 -DHAVE_LOCALE -DHAVE_STRTOF_L`

The branch is here fwiw: https://github.com/endrift/mgba/tree/feature/wasm

BinBashBanana commented:
Does that go in the makefile in mGBA or in RetroArch?

endrift commented:
Makefile.libretro line 390, you added a line right after it.

BinBashBanana commented:
I reverted formatting.c and formatting.h, added -DHAVE_STRTOF_L to Makefile.libretro, tried recompiling but got this error:
```
src/util/formatting.c:67:14: error: implicit declaration of function 'strtof_l' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        float res = strtof_l(str, end, l);
                    ^
src/util/formatting.c:67:14: note: did you mean 'strtof_u'?
src/util/formatting.c:61:7: note: 'strtof_u' declared here
float strtof_u(const char* restrict str, char** restrict end) {
      ^
1 error generated.
```

endrift commented:
You should try also adding -DHAVE_XLOCALE=1

BinBashBanana commented:
It throws the same error.

endrift commented:
Try `-D_GNU_SOURCE` instead of `-DHAVE_XLOCALE`

endrift commented:
Looks good now, thanks

BinBashBanana commented:
You beat me to it. :)
